core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala

```diff
@@ -175,6 +175,7 @@ private[spark] class IndexShuffleBlockResolver(
           dataFile.delete()
         }
         if (!indexTmp.renameTo(indexFile)) {
+          indexTmp.delete()
```

Comment:
To be maximally paranoid and safe I should add a similar `finally` block here to enforce the lifetime of the `indexTmp` file.